server/src/main/java/org/opensearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -479,74 +483,97 @@ public void cloneShardSnapshot(
-                newGen = String.valueOf(tuple.v2() + 1);
-                existingSnapshots = tuple.v1();
             } else {
                 newGen = UUIDs.randomBase64UUID();
-                existingSnapshots = buildBlobStoreIndexShardSnapshots(Collections.emptySet(), shardContainer, shardGeneration).v1();
-                existingShardGen = shardGeneration;
-            }
-            SnapshotFiles existingTargetFiles = null;
-            SnapshotFiles sourceFiles = null;
-            for (SnapshotFiles existingSnapshot : existingSnapshots) {
-                final String snapshotName = existingSnapshot.snapshot();
-                if (snapshotName.equals(target.getName())) {
-                    existingTargetFiles = existingSnapshot;
-                } else if (snapshotName.equals(source.getName())) {
-                    sourceFiles = existingSnapshot;
-                }
-                if (sourceFiles != null && existingTargetFiles != null) {
-                    break;
-                }
-            }
-            if (sourceFiles == null) {
-                throw new RepositoryException(
-                    metadata.name(),
-                    "Can't create clone of ["
-                        + shardId
-                        + "] for snapshot ["
-                        + target
-                        + "]. The source snapshot ["
-                        + source
-                        + "] was not found in the shard metadata."
+                RemoteStoreShardShallowCopySnapshot remStoreBasedShardMetadata = loadRemStoreEnabledShardSnapshot(shardContainer, source);
+                REMOTE_STORE_SHARD_SHALLOW_COPY_SNAPSHOT_FORMAT.write(
+                    remStoreBasedShardMetadata.asClone(target.getName(), startTime, threadPool.absoluteTimeInMillis() - startTime),
+                    shardContainer,
+                    target.getUUID(),
+                    compress
                 );
-            }
-            if (existingTargetFiles != null) {
-                if (existingTargetFiles.isSame(sourceFiles)) {
-                    return existingShardGen;
-                }
-                throw new RepositoryException(
-                    metadata.name(),
-                    "Can't create clone of ["
-                        + shardId
-                        + "] for snapshot ["
-                        + target
-                        + "]. A snapshot by that name already exists for this shard."
+                String indexUUID = remStoreBasedShardMetadata.getIndexUUID();
+                String remoteStoreRepository = remStoreBasedShardMetadata.getRemoteStoreRepository();
+                RemoteStoreMetadataLockManager remoteStoreMetadataLockManger = remoteStoreLockManagerFactory.newLockManager(
+                    remoteStoreRepository,
+                    indexUUID,
+                    String.valueOf(shardId.shardId())
+                );
+
+                remoteStoreMetadataLockManger.cloneLock(
```

Comment:
Lock should be cloned first and then the file `REMOTE_STORE_SHARD_SHALLOW_COPY_SNAPSHOT_FORMAT` should be created.

Also what is the behavior if there is failure after creating the lock file but before writing the metadata file ? Assuming even if the operation fails, how is the leaked lock file deleted ?

Reply:
Handled the failure in new revision. If clone operations fails for a shard, we delete the created lock file.